---
title: "Gene differential expression analysis"
author: "Chunhui"
date: "7/8/2019"
output:
  html_document:
    toc: yes
  pdf_document:
    toc: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## PART 1: Preparation: Packages & options

```{r, include=FALSE}
# this option assigns more RAM for java to enable high-RAM-comsuming processing
# especially in writing large dataset to .xlxs
rm(list = ls())
memory.limit(102400)
options(java.parameters = "-Xmx10000m")

```

if you have problem about 'could not find function', make sure you got all the library first
```{r packages, include=FALSE}
library(BiocManager)
# if you have install DESeq2, uncomment the followiing line
# BiocManager::install("DESeq2")
library(DESeq2)
library(sqldf)
library(tidyverse)
library(biomaRt)
library(dplyr)
library(xlsx) 
library(pheatmap) #for heatmap
library(gplots) #for heatmap
# run next line if you need to get some help from DESeq2
# browseVignettes("DESeq2")
```


## PART2: Data preparation

### 2.1 reading dataset from csv file
```{r}
head(table <- read.csv('kinetics_All_HD.csv', stringsAsFactors = F))
```
The first column is the `ensemble gene id`, the second column is a mixture of `ensemble gene id` and `external gene name`.
The third column to the last column are gene expression level, the prefix indicates which patient we get the 
sample and the suffix indicates which time point we get the sample.
In this example, there are total 3 patients and each patient has 4 time samples, 12 total.


### 2.2 Get external
For further analysis, we need to get `external gene name` from `ensenble gene id` as possible as we can. By nature, `external gene name` and `ensenble gene id` are not one-to-one relationship, e.g. multiple `external gene names` could correspond to one `external gene name`

To do this, we need library `BiomaRt`.

List available `Mart` by use `listMarts`, then create ensembl objection by using `useMart`.
Use `listDatasets` with ensembl objection as parameter to get available gene dataset

```{r eval=FALSE, message=FALSE, include=FALSE}
listMarts()
ensembl=useMart("ensembl")
listDatasets(ensembl)
```

Create Mart object by using `useDataset` with corresponding `ensemble` and `gene dataset`. 
`listFilters` would list all the available filters

For more information about `biomaRt`, run **browseVignettes('biomaRt')**
```{r eval=FALSE, include=FALSE}

mart <- useEnsembl(biomart = "ensembl", 
                   dataset = "hsapiens_gene_ensembl",
                   mirror = 'useast'
                   )
listFilters(mart)
```

use `getBM` with correct filter and attributes to get corresponding `external gene name` from ` ensemble gene id`

**This procedure could take a long time depending how large your dataset and you network speed**
```{r eval=FALSE, include=FALSE}
G_list_2 <- getBM(filters= "ensembl_gene_id", 
                  attributes= c("ensembl_gene_id", 'external_gene_name'),
                  values=table$Geneid1,
                  mart = mart,
                  )
head(G_list_2)
```

```{r}
G_list_2 <- read.xlsx2(file = 'G_list.xlsx', sheetIndex = 1)
```


## rebuild dataset
left join original dataset `table` with `G_list`, for those `ensembl gene name` don't have `external gene name` , using their `ensemble gene name` instead.

```{r echo=TRUE}
table_merge <- sqldf::sqldf('select *
             from "table"
             left join G_list_2 on "table".Geneid1 = G_list_2.ensembl_gene_id') 
table_reordered <- table_merge %>%
  as_tibble() %>% 
   dplyr::select(Geneid1, external_gene_name, everything())
#remove useless column
table_reordered <- table_reordered[, c(-3, -19)]
table_reordered$external_gene_name <- 
  ifelse(is.na(table_reordered$external_gene_name),table_reordered$Geneid1,
         table_reordered$external_gene_name)

head(table_reordered)
```

keep only `Geneid` column
```{r}
table_geneid <- table_reordered[,-1]
colnames(table_geneid)[1] = 'Geneid'
# omit any row if has missing value 
# Caution! this may not the best method
head(table_geneid)
```

## resolve duplicates 
Since the transformation from `ensemble gene id` and `external gene name` is not one-to-one, so `Geneid` is not unique, to transform dataset into a matrix with row name, `Geneid` must be `unique`
** for duplicates `Geneid`, take the mean of each column into one row**

```{r}
#remove depublicate by geneid using sqlite
# base rule: average the read from the same gene
# caution! This may not applicable for bio-duplicate
table_nodup <-  sqldf::sqldf('
  select Geneid,
         avg(HD1_Blood) as HD1_Blood, avg(HD2_Blood) as HD2_Blood, avg(HD6_Blood) as HD6_Blood,
         avg(HD1_1H) as HD1_1H, avg(HD2_1H) as HD2_1H, avg(HD6_1H) as HD6_1H,
         avg(HD1_2H) as HD1_2H, avg(HD2_2H) as HD2_2H, avg(HD6_2H) as HD6_2H,
         avg(HD1_4H) as HD1_4H, avg(HD2_4H) as HD2_4H, avg(HD6_4H) as HD6_4H,
         avg(HD1_6H) as HD1_6H, avg(HD2_6H) as HD2_6H, avg(HD6_6H) as HD6_6H
  from table_geneid
  group by Geneid
  order by Geneid
  '
)
head(table_nodup)
```

Double check whether there is no duplicates
```{r}
#check how many gene left in the processed dataset
length(table_nodup$Geneid)

#check whether there is duplicate now, if there is no duplicate, a True should be returen
all(!duplicated(table_nodup[,1])) 
```

## PART 3: Using DESeq2
`DESque2` is the updated version of `DESque`,
You may have different type of orginal file, `DESeq2` has different data preparation solutions for them.
For more information,m run **browseVignettes('DESeq')**

### 3.1 Data preparation
make the matrix needed for next step


```{r}
#make the matrix needed for next step
#use the geneid as row.names 
table_rownames <- data.frame(table_nodup[,-1], row.names=table_nodup[,1])
count_maxtrix <- as.matrix(table_rownames)
# make the mode of matrix integer otherwise it will be number
mode(count_maxtrix) <- 'integer'
head(count_maxtrix)
```

```{r}

#create the coldata corresponding to the processed dataset
coldata <- data.frame(condition = c(rep('blood', 3),
                                    rep('1h', 3),
                                    rep('2h', 3),
                                    rep('4h', 3),
                                    rep('6h', 3)),
                      row.names = colnames(count_maxtrix)[1:15])
head(coldata)

```

It is absolutely critical that the columns of the count matrix and the rows of the column data (information about samples) are in the same order. DESeq2 will not make guesses as to which column of the count matrix belongs to which row of the column data, these must be provided to DESeq2 already in consistent order.

```{r}
#check whether the name of row match col, a True should be returned
all(rownames(coldata) == colnames(count_maxtrix)[1:15])

#create DESeqDataSet(dds) object, dds is a container for intermediate data
dds <- DESeqDataSetFromMatrix(countData = count_maxtrix[,1:15],
                              colData = coldata,
                              design = ~ condition)
```

### 3.2 pre-filtering
```{r}
# pre-filtering
# by removing rows in which there are very few reads, we reduce the memory size of the dds data object, 
#   and we increase the speed of the transformation and testing functions within DESeq2
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]

# set dds condition with all time points
# set 'blood' as base level, the default base level is determined by alphabet order
# other statement if also available for this purpose
# Caution!: the document only give example for factor with two levels, not sure about accuracy for 
#   factor with more than 2 levels
dds$condition <- factor(dds$condition, levels = c("blood","1h",'2h', '4h', '6h'))
# drop levels that with no sample
dds$condition <- droplevels(dds$condition)
```


## PArt 4: Defferential Expression Analysis

### 4.1 Create dds object
```{r}
# DEseq analysis
dds <- DESeq(dds)
# for different condition you just change 6h to i.e. 1h, 2h, 4h
# may can be done by lapply() function, not sure how to set the func argument
# with different arguments
res_1h <- results(dds,contrast=c("condition","1h","blood"))
res_2h <- results(dds,contrast=c("condition","2h","blood"))
res_4h <- results(dds,contrast=c("condition","4h","blood"))
res_6h <- results(dds,contrast=c("condition","6h","blood"))
resultsNames(dds)
```

### 4.2 Build Function for subset result
`res_subgroup` is a useful function to subset the result from DESeq, it has 4 parameter
`res` is an instance of DESeq.result object
`alpha` is the boundary for adjusted p-value, default 0.1 if not specified.
`reg_dir` is the direction of your `alpha` value, 'all', 'up' and 'down' can be used.
  all  -- up and down regulated 
  up   -- up regulated only
  down -- down regulated only
```{r subset result}
# this function is used to subset the neeeded gene from RESULT object
#function get those LFC greate than 1 and the adjusted P-value is less than 0.1
res_subgroup <- function(res, alpha=0.1, reg_LFC=1, reg_dir='all'){
  
  # res is an obj from DEseq2.result function 
  # alpha gives the significant level for adjusted P-value
  # reg gives the regulation level change in log2 fold change in absolute value
  # reg_dir gives which regulation direction you want to subset you gene
    # three options: all -- up and down 
    #                up  -- only up regulated
    #                down -- only down regulated
  res_sig_pos <- (res$padj < alpha)
  res_sig_pos[is.na(res_sig_pos)] <- F
  if(reg_dir == 'all'){
    res_LFC_pos <- (res$log2FoldChange > reg_LFC) | (res$log2FoldChange < -reg_LFC)
  }
  else if(reg_dir == 'up'){
           res_LFC_pos <- (res$log2FoldChange > reg_LFC)
       }
       else if(reg_dir == 'down'){
                res_LFC_pos <- (res$log2FoldChange < -reg_LFC)
            }
  
  return(res[res_LFC_pos & res_sig_pos,])
  
}

```

You can use `lapply` to apply a function to a list of dataset togather of different `condition`, e.g. 1h_vs_blood, 2h_vs_blood, 4h_vs_blood and 6h_vs_blood
```{r}
#get all 4 upregluated genes
res_list <- list(res_1h,res_2h,res_4h,res_6h)
res_list_up <- lapply(res_list, res_subgroup, reg_dir ='up')
#get all 4 downregluated genes
res_list_down <- lapply(res_list, res_subgroup, reg_dir ='down')
```

This code block is used for easily output xlxs file with mnemonic name.
```{r output xlsx file, eval=FALSE, include=FALSE}
# output corresponding excel file
library(xlsx)
# genes that are upregulated
file_up <- paste(getwd(),'/', c("1", '2', '4', '6'), '_up.xlsx', sep="")
for(i in 1:4){
  write.xlsx2(res_list_up[i], file = file_up[i], row.names = T)
}
# genes that are downregulated
file_down <- paste(getwd(),'/', c("1", '2', '4', '6'), '_down.xlsx', sep="")
for(i in 1:4){
  write.xlsx2(res_list_down[i], file = file_down[i], row.names = T)
}
```


### A-not-B subset with sql
You can use code below to subset genes that only in A `res_LTB4_up` and not in B `res_CFASN_up`

```{r eval=FALSE, include=FALSE}
res_LTB4_up.unique <- sqldf::sqldf('
              select *
              from res_LTB4_up
              where res_LTB4_up.gene_name not in (
                select res_LTB4_up.gene_name
                from res_LTB4_up
                inner join res_CFASN_up on
                  res_LTB4_up.gene_name = res_CFASN_up.gene_name
              )
             ')
```


## PART 5: Analysis and plot


###  5.1 Plot MA-plot
use `plotMA` function with a `result` object as parameter 
```{r}
# plot the MAplot with the subset result
plotMA(res_1h, ylim=c(-3,3))
plotMA(res_list_up[[1]], ylim=c(-3,3))
```


### 5.2 Correlation 
This is a demo for correlation gene in a list `overlap transcriptome proteome.xlsx` of **expression of proteomics**(`Proteomics.xlsx`) and **expression of RNAseq**(`10hr_normed_counts.xlsx`). 

First we need to get the RNAseq and proteomics level from two data files.
```{r}
# correlation analysis
#read in data as tibble
proteomics <- read.xlsx2(file = 'Proteomics.xlsx', sheetName = 'Sheet1', stringsAsFactors = F,
                         colClasses = c('character', 'character', rep("double", 9))
                         ) %>% 
  as_tibble()

count_10h <- read.xlsx2(file = '10hr_normed_counts.xlsx', sheetName = 'counts',
                        colClasses = c('character', rep("double", 16))) %>% 
  as_tibble()

# delete the '_HUMAN' suffix in the GENE.ID variable of proteomics dataset.
proteomics$GENE.ID <- sapply(proteomics$GENE.ID, sub, pattern = '_HUMAN', replacement = '')

# 
intersc <- intersect(proteomics$GENE.ID, count_10h$ID)
x <- count_10h[count_10h$ID %in% intersc, ] %>% 
  dplyr::select(ID,CFASN_TM_1:CFASN_TM_3)
y <- proteomics[proteomics$GENE.ID %in% intersc,] %>% 
  dplyr::select(GENE.ID,CFASN_1:CFASN_3)

```

Read in the list of genes need to analyze correlation.
```{R}
intersc.up <- as.list(read.xlsx2('overlap transcriptome proteome.xlsx', header = F, 
                                 sheetIndex = 1, stringsAsFactors = F))

```

```{r warning=FALSE}
##get the correaltions of transcriptomics and proteinomics of intersection genes
# Initiate data frame of correlation results
correlations <- data.frame(matrix(NA, nrow=length(x$ID), ncol=6))
# Assign the first column as the protein names and their site (B2/B4) subset
correlations[ ,1] <- x[,1]
# Assign column names
colnames(correlations) <- c("gene", "Sample Size", "Pearson CC", "Pearson P-Value",
                            "Spearman Rho", "Spearman P-Value")

for (i in 1:nrow(correlations)) {
  cor.pearson <- cor.test(as.numeric(x[i,][,2:4]), as.numeric(y[i,][,2:4]))
  cor.spearman <- cor.test(as.numeric(x[i,][,2:4]), as.numeric(y[i,][,2:4]), method = 'spearman')
  #Col 2. Sample size
  correlations[i,2] <- as.integer(cor.pearson$parameter) + 2
  #Col 3. pcc
  correlations[i,3] <- as.numeric(cor.pearson$estimate)
  #Col 4. pcc p-value
  correlations[i,4] <- as.numeric(cor.pearson$p.value)
  #Col 6. spearman's rho
  correlations[i,5] <- as.numeric(cor.spearman$estimate )
  #Col 7. spearman's rho p-value
  correlations[i,6] <- as.numeric(cor.spearman$p.value )
}

```

Due to the small sample size, most of points has a large P-value, you can use the code commented to get those with p-value < 0.05
```{r}
#correlations$`Pearson CC`[which(correlations$`Pearson P-Value` < 0.05)]
plot(correlations$`Pearson CC`, pch = 16, cex = 0.5)
```



### 5.3 plot trend through time
First get the data we need
```{r}
library(ggplot2)
library(tidyr)
dat <- NULL
dat$time <- as.factor(c('1h', '2h', '4h', '6h'))
#get the number of genes which have significant differential expression
dat$down <- sapply(res_list_down, nrow)
dat$up <- sapply(res_list_up, nrow)
dat <- as.data.frame(dat)
```

The plot show the there is a huge increase of # of down-regulated genes  
```{r}
# gather function is useful to takes multiple columns and collapses into key-value pairs
ggplot(data = dat %>% gather(Variable, reg, -time), 
       aes(x = time, y = reg, fill = Variable)) + 
  geom_bar(stat = 'identity', position = position_dodge(width = 0.15)) +
  geom_line(aes(x = time, y = reg, group = Variable, color = Variable),
            stat="identity", show.legend = F) +
  geom_point(show.legend = F) +
  scale_fill_manual(values=c("black", "grey")) +
  labs(title = "Down regulated genes over time", x = "time", y = "DEGs")
```


### 5.4 Bubble plot
Bubble plot for gene enrichment analysis.
Data preparation for the bubble plot.
```{r}
data.tf.enrich <- read.xlsx2(file = 'Enrichment figures R Chunhui.xlsx',
                                   sheetName = 'TF',
                                   colClasses = c('character','numeric','character', rep('numeric', 4))
)

head(data.tf.enrich)
```
Your enrichment data may looks like this.


```{R}
data.tf.enrich$Gene.Set.Name <- paste('                               ', data.tf.enrich$Gene.Set.Name) 
data.tf.enrich$FDR.q.value = -log10(as.numeric(data.tf.enrich$FDR.q.value))
#order the data by # of genes overlaped
data.tf.enrich$Gene.Set.Name <- factor(data.tf.enrich$Gene.Set.Name,                                              levels=data.tf.enrich$Gene.Set.Name[order(data.tf.enrich$X..Genes.in.Overlap..k.)])
```

The color indicates `FDR q value`.
The size of circle indicates `# of genes overlaped`.
```{r}
ggplot(data = data.tf.enrich) + 
  geom_point(aes(x = data.tf.enrich$Gene.Set.Name, y='', size =  X..Genes.in.Overlap..k., color = FDR.q.value)) +
  coord_flip() + 
  labs(title = 'TF', size = '# of Genes in Overlap (k)', color = 'FDR q value') + 
  ylab('') +
  xlab('Genes') +
  scale_colour_gradient(low = "red", high = "green") +
  scale_size(breaks = c(50, 100, 150, 200, 250)) +
  theme(legend.position = "right") 
  
```

# 5.5 heatmap
Heatmap for a list of genes `heatmap CF upregulated.xlsx` among the four time points
First, get the matrix for heatmap.
```{r}
gene_intsc.table <- read.xlsx2(file = 'heatmap CF upregulated.xlsx', sheetIndex = 1,
                               stringsAsFactors = F, header = F)

lfc_ts_expr.tibble <- tibble(gene_name = rownames(res_1h),
      lfc.1h = res_1h$log2FoldChange,
      lfc.2h = res_2h$log2FoldChange,
      lfc.4h = res_4h$log2FoldChange,
      lfc.6h =res_6h$log2FoldChange)

#filter genes and create matrix with column names
gene_intsc_heatmap.tibble <- lfc_ts_expr.tibble %>%
  dplyr::filter(gene_name %in% gene_intsc.table[,1])
gene_intsc_heatmap.matrix <- as.matrix(gene_intsc_heatmap.tibble[,2:5])

# set the row names for matrix
rownames(gene_intsc_heatmap.matrix) <- as.vector(as.list(gene_intsc_heatmap.tibble[,1])[[1]])
```

**You may need to zoom or set resolution in export to see gene name clearly**
```{R}
pheatmap(gene_intsc_heatmap.matrix, cluster_cols = F)
```
`pheatmap` and `heatmap.2` are two function available to plot heatmap.
see package `pheatmap` and `gplots` for more details.
```{r}
# Since we don't need column cluster, we set Colv = FALSE and dendrogram = 'row'
heatmap.2(gene_intsc_heatmap.matrix, Colv = FALSE, dendrogram = 'row')
```



Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
